lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU

```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+        FOR i := (udBackgroundVoltageBufferIndex - 5) TO udBackgroundVoltageBufferIndex DO
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageBuffer[i MOD 100];
+        END_FOR
+        udBackgroundVoltageBufferIndex := udBackgroundVoltageBufferIndex - 5;
+        IF udBackgroundVoltageBufferIndex <> 1 AND BACKGROUND_MODE = Passive THEN
+            fBackgroundVoltage := fBackgroundVoltageSum / MAX((udBackgroundVoltageBufferIndex - 1), 100);
+        END_IF
+    END_IF
+    fBackgroundVoltageAcc := 0;
+    fBackgroundVoltageSum := 0;
+    tonBackgroundAutoCollecting(IN := FALSE);
+    udBackgroundVoltageBufferIndex := 1;
+ELSIF eEnumGet = E_PPM_States.OUT AND stYAxisState = 0 THEN
+    tonBackgroundAutoCollecting( IN := TRUE, PT := T#1000MS);
+    IF tonBackgroundAutoCollecting.Q = TRUE THEN
+        tonBackgroundAutoCollecting( IN := FALSE);
+        fBackgroundVoltageStale := fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex];
+        fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex] := fBackgroundVoltageAcc;
+        IF udBackgroundVoltageBufferIndex > 100 THEN
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageStale;
+        END_IF
+        fBackgroundVoltageSum := fBackgroundVoltageSum + fBackgroundVoltageAcc;
+        fBackgroundVoltageAcc := fVoltage;
+        udBackgroundVoltageBufferIndex := udBackgroundVoltageBufferIndex + 1;
+    ELSE
+        fBackgroundVoltageAcc := fBackgroundVoltageAcc + fVoltage;
+    END_IF
+
+END_IF
+
+IF udBackgroundVoltageBufferIndex <> 1 AND BACKGROUND_MODE = Passive THEN
+    fBackgroundVoltage := fBackgroundVoltageSum / MAX((udBackgroundVoltageBufferIndex - 1), 100);
+END_IF
+
+IF bBackgroundCollect = TRUE THEN
+    bBackgroundCollect := FALSE;
+    BACKGROUND_MODE := Manual;
+    udBackgroundManualCount := 0;
+    fBackgroundVoltage := 0;
+    tofBackgroundCollecting(IN := TRUE);
+END_IF
+
+IF BACKGROUND_MODE = Manual THEN
+    tofBackgroundCollecting(IN := FALSE, PT := UINT_TO_TIME(uBackgroundCollectionTime * 1000));
+    IF tofBackgroundCollecting.Q = TRUE THEN
+        udBackgroundManualCount := udBackgroundManualCount + 1;
+        fBackgroundVoltage := (fBackgroundVoltage * (udBackgroundManualCount - 1) + fVoltage) / udBackgroundManualCount;
+    END_IF
+END_IF
+
+// Getting frequency based on accelerator source
+CASE uAccelerator OF
+    0: uFrequency := uNCFrequency;
+    1: uFrequency := uSCFrequency;
+    2: uFrequency := 0; // ASTA?
+ELSE
+    uFrequency := 0;
+END_CASE
+
+// Getting wattage and energy
+fPulseWattage := (fVoltage - fBackgroundVoltage) / fResponsivity;
+IF fResponsivity <> 0 AND uFrequency <> 0 THEN
+    fCalibMJ := fPulseWattage / uFrequency;
+ELSE
+    fCalibMJ := -9999;
+END_IF
+
+CASE PMPS_GVL.cstFullBeam.neV OF
+    0..300: 				bOverAllowableEnergy := fCalibMJ > 2;
+    fWavelengthEdge1..1000: bOverAllowableEnergy := fCalibMJ > 4;
+    fWavelengthEdge2..1600: bOverAllowableEnergy := fCalibMJ > 0.5;
```

Comment:
The lower edges here are defined as constants above, but the upper edges are hard-coded here. There's probably a way to implement this more consistently.
Something like (details/syntax issues notwithstanding)
```
0..fWaveLengthEdge1
fWavelengthEdge1+1...fWavelengthEdge2
```